proposals/NNNN-package-access-modifier.md

```diff
@@ -0,0 +1,326 @@
+
+```
+@usableFromPackageInline func internalFuncA() {}
+@usableFromInline func internalFuncB() {}
+@usableFromInline package func packageFunc() {}
+
+@inlinable package func pkgUse() {
+    internalFuncA() // OK
+    internalFuncB() // OK
+    packageFunc() // OK
+}
+
+@inlinable public func publicUse() {
+    internalFuncA() // ERROR
+    internalFuncB() // OK
+    packageFunc() // OK
+}
+```
+ 
+### Resiliency
+
+Library evolution makes modules resilient.  We can incorporate a package name into a resiliency check and bypass it if modules are in the same package.  This will remove the need for resilience overhead such as indirection and language requirements such as `@unknown default` for an unfrozen `enum`.  
+
+### Subclassing and Overrides
+
+Access control in Swift usually doesn't distinguish between different kinds of use.  If a program has access to a type, for example, that gives the programmer a broad set of privileges: the type name can be used in most places, values of the type can be borrowed, copied, and destroyed, members of the type can be accessed (up to the limits of their own access control), and so on.  This is because access control is a tool for enforcing encapsulation and allowing the future evolution of code.  Broad privileges are granted because restricting them more precisely usually doesn't serve that goal.
+
+However, there are two exceptions.  The first is that Swift allows `var` and `subscript` to restrict mutating accesses more tightly than read-only accesses; this is done by writing a separate access modifier for the setter, e.g. `private(set)`.  The second is that Swift allows classes and class members to restrict subclassing and overriding more tightly than normal references; this is done by writing `public` instead of `open`.  Allowing these privileges to be separately restricted serves the goals of promoting encapsulation and evolution.
+
+Because setter access levels are controlled by writing a separate modifier from the primary access, the syntax naturally extends to allow `package(set)`.  However, subclassing and overriding are controlled by choosing a specific keyword (`public` or `open`) as the primary access modifier, so the syntax does not extend to `package` the same way.  This proposal has to decide what `package` by itself means for classes and class members.  It also has to decide whether to support the options not covered by `package` alone or to leave them as a possible future direction.
+
+Here is a matrix showing where symbols with each current access level can be used or overridden:
+
+<table>
+<thead>
+<tr>
+<th>Subclassable in... \ Accessible in...</th>
```

Comment:
Hmm.  I get what the slash is aiming to do, but I think it would be easier to read to just make the row labels be "Subclassable anywhere" etc. and the column labels "Accessible anywhere" etc.  Alternatively, you could put the prefixes in cells that use `rowspan` and `colspan`, if the markdown processing doesn't mess those up.